### ISSUE TITLE
replace all usage of pywin32 with cffi

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#7477]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)| @glyph reported|
|-|-|
|Trac ID|trac#7477|
|Type|enhancement|
|Created|2014-06-05 03:48:50Z|

Getting Python and the relevant development tools working on Windows is a big pain.

pywin32 really exacerbates this problem because it is resolutely stuck in a troubled period of Python's past:

  1. It's not present as a download in any form on PyPI, so `pip install` and `easy_install` don't work.
  2. They don't ship binary wheels, so installation takes a long time even if you've got a fully set up development environment.
  3. The EXE installers have [some weird incompatibilities](https://stackoverflow.com/questions/8078216/how-to-install-pywin32-on-virtualenv-with-python-2-7-64-bit) which means that you can't even cram them into a virtualenv with easy_install.
  4. Want to fix this to build your own wheels to work around this?  If you don't have MSDN, [too bad, you can't build it with the free tools and they aren't going to fix that](http://sourceforge.net/p/pywin32/bugs/520/).  Maybe too bad even if you do, I haven't tried with a full install of VS2K8 Pro yet.

Given that we already depend on cffi, and all we're using pywin32 for is to call a couple of win32 APIs, it seems like we could drop the dependency entirely without an unreasonable amount of effort.

<details><summary>Searchable metadata</summary>

```
trac-id__7477 7477
type__enhancement enhancement
reporter__glyph glyph
priority__normal normal
milestone__ 
branch__ 
branch_author__ 
status__assigned assigned
resolution__None None
component__core core
keywords__ 
time__1401940130000000 1401940130000000
changetime__1560427391945218 1560427391945218
version__None None
owner__opalmer opalmer
cc__adiroiban cc__frederic.wang cc__cas--
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_1"></a>|@glyph commented|
|-|-|

For what it's worth, I have managed to build a binary wheel with just a couple of tweaks (changing `os.environ["PATH"]` in a pth file so that some DLLs that the PYDs depend on can be found).  [You can find it here](https://glyph.im/pywin32/index/pywin32/).

**Comment by twisted-trac** [comment]
|<img alt="tritium21's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_2"></a>|tritium21 commented|
|-|-|

To point 4: that isn't exactly accurate.  You *CAN* build it with free tools, VS2K8Express is just not the tool to use IF, and ONLY IF you are trying to build 64bit binaries.  You CAN use the compiler that comes with the windows 7 SDK.  This ONLY effects mainline python builds of 2.6 - 3.2.  3.3 is built with 2K10, the express version of which comes with a 64 bit compiler.  In order for them to fix the issue that 'they don't care about' is for Mark Hammond to buy Microsoft.

In other words, this is not a pywin32 issue, its a Microsoft issue, and only for the minority of users who use 64 bit versions of mainline Python.

EDIT:
Ok, I have to redact part of this.  I assumed, and everyone knows what that is an acronym for.

Anyways, mc.exe comes with the windows 7 sdk, which is the full compiler suit for windows, and it is free.  so the main point still stands, that there are free tools that will compile this ...increasingly hunk of junk module... but the 2008 express editions are not one of those tools.

**Comment by twisted-trac** [comment]
|[<img alt="zooko's avatar" src="https://avatars.githubusercontent.com/u/467900?s=50" width="50" height="50">](https://github.com/zooko)<a name="note_3"></a>|@zooko commented|
|-|-|

FWIW, replacing pywin32 with cffi is the approach we took in Tahoe-LAFS. We had quite a few issues back when Tahoe-LAFS used pywin32:

* https://tahoe-lafs.org/trac/tahoe-lafs/ticket/137 "automate satisfaction of pywin32 dependency"
* https://tahoe-lafs.org/trac/tahoe-lafs/ticket/142 "pywin32 can't be installed with setuptools"
* https://tahoe-lafs.org/trac/tahoe-lafs/ticket/756 "if pywin32 has been manually installed, setuptools still doesn't detect it"

(Probably more tickets exist, but you get the idea.)

Then we finally removed our dependency on pywin32 altogether:

* https://tahoe-lafs.org/trac/tahoe-lafs/ticket/1274 "eliminate pywin32 dependency"

You might be interested to look at how we did it, by using ctypes to invoke the Windows native APIs and by avoiding Twisted's subprocess features in favor of the builtin "subprocess" module:

* https://tahoe-lafs.org/trac/tahoe-lafs/changeset/d21f4071c3229e18/trunk
* https://tahoe-lafs.org/trac/tahoe-lafs/changeset/6dd8b6f47126fdc0/trunk

We were subsequently pretty disappointed when a new version of Twisted came out which added a hard dependency on pywin32, thus undoing some of the benefits we had gained. :-)

* https://tahoe-lafs.org/trac/tahoe-lafs/ticket/2028 "Twisted endpoints introduce a dependency on pywin32"

Our latest hackey idea is to switch from pywin32 to pypiwin32:

* https://tahoe-lafs.org/trac/tahoe-lafs/ticket/2392 "add dependency on pypiwin32 to satisfy Twisted on Windows"

But if a new version of Twisted came out that did everything on Windows without using pywin32, that would be great for us!

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_4"></a>|@glyph commented|
|-|-|

pypiwin32 is definitely a temporary workaround while this can be done.  We really don't touch that much of the API surface of pywin32.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_5"></a>|@glyph commented|
|-|-|

Instead of just using the gross URL in the first comment on this ticket, you can, however, now [https://warehouse.python.org/project/pypiwin32 "`pip install pypiwin32`"] and the new [ticket:3696 extras support] includes a reference to it.

**Comment by twisted-trac** [comment]
|<img alt="opalmer's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_6"></a>|opalmer set owner to opalmer<br>opalmer set status to `assigned`|
|-|-|

Going to go ahead and assign this to myself since I've started on this and have at least one module (python.lockfile) clear of pywin32.  There's still a lot of work to be done with this so I'll be keeping this branch on github up to date as I work through the rest of the library:

https://github.com/twisted/twisted/compare/trunk...opalmer:windows-cffi

It's a bit of a mess right now but hopefully within the week or so I can get it cleaned up and working properly with the setup.py.  Once that's done I might create an initial patch set for review.  That might keep the reviews smaller and also make this a bit more incremental (unless someone has a better suggestion).

**Comment by twisted-trac** [comment]
|<img alt="opalmer's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_7"></a>|opalmer commented|
|-|-|

Broke off the first change into [#7889](https://github.com/twisted/twisted/issues/7889).  The branch in the link above was getting quite large so I've opted to start a little smaller first and replace a couple of functions that hopefully will be easier to test and review.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_8"></a>|@adiroiban commented|
|-|-|

How is Twisted already dependent on cffi?

I see that [#6831](https://github.com/twisted/twisted/issues/6831) is not merged yet.

Maybe we need a separate ticket to add cffi to dist.py and document that for Twisted it is planned to replace ctype usage with cffi

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_9"></a>|@glyph commented|
|-|-|

Replying to [adiroiban](#note_8):
> How is Twisted already dependent on cffi?

It has no hard dependency on cffi yet.  But you need cffi to use any of its TLS stuff.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_10"></a>|@glyph commented|
|-|-|

From frederic.wang (spam-filtered, unfortunately):

----------

Any update on this?

(I want to run buildbot slaves on MSYS2 shells and cffi seems much easier to build with mingw64 than pywin32)

**Comment by twisted-trac** [comment]
|<img alt="frederic.wang's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_11"></a>|frederic.wang commented|
|-|-|

Resaying what I wrote on [#7889](https://github.com/twisted/twisted/issues/7889), although I'll probably be spam-filtered again. Enthought started to rewrite pywin32 with ctypes/cffi:

https://github.com/enthought/pywin32-ctypes
https://pypi.python.org/pypi/pywin32-ctypes/0.0.1

"The default behaviour will try to use cffi if available and fall back to using ctypes."

The API seems very limited at the moment.

**Comment by twisted-trac** [comment]
|[<img alt="cas--'s avatar" src="https://avatars.githubusercontent.com/u/606038?s=50" width="50" height="50">](https://github.com/cas--)<a name="note_12"></a>|@cas-- commented|
|-|-|

This is a gentle nudge to find out if this can be revived. 

For the Deluge bittorrent client we use GTK and Twisted and packaging for Windows has got much harder recently with trying to move to GTK3 where the recommended install method is via MinGW. However Twisted is [not available](https://github.com/Alexpux/MINGW-packages/issues/893), primarily due to pywin32. We also use [pywin32](https://dev.deluge-torrent.org/ticket/3181) but since Twisted depends on it, I've delayed migrating away from it.

I found that Oliver Palmer has created a new module [pywincffi](https://github.com/opalmer/pywincffi) but it seems that development has stalled, although there is also a very rough PR based upon that work: https://github.com/twisted/twisted/pull/920

Any chance new progress could be made on this?

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_13"></a>|@glyph commented|
|-|-|



Replying to [Calum Lind](#note_12):

Hi Calum!  Thanks for chiming in.  Deluge is definitely an important Twisted user (hey, have you ever considered contributing to https://twistedmatrix.com/trac/wiki/SuccessStories?) and it would be great to get this fixed.

> This is a gentle nudge to find out if this can be revived. 

Absolutely!  You may also be interested in [#7945](https://github.com/twisted/twisted/issues/7945), which suggests putting all of this stuff in _optional_, rather than required, dependencies.

> For the Deluge bittorrent client we use GTK and Twisted and packaging for Windows has got much harder recently with trying to move to GTK3 where the recommended install method is via MinGW. However Twisted is [not available](https://github.com/Alexpux/MINGW-packages/issues/893), primarily due to pywin32.

Is this because pywin32 can't be built against minGW's libraries or C runtime or something?  More detail would be helpful.  We stopped depending against my "pypiwin32" fork because pywin32 _did_ start building binary wheels, which is supposed to solve this whole mess.

Looking at https://pypi.org/project/pywin32/#files right now I see wheels for 32bit / 64bit x 27 35 36 37, which seems like a reasonable collection.  Sad face, no PyPy, but it doesn't sound like that's the problem you're having.

> We also use [pywin32](https://dev.deluge-torrent.org/ticket/3181) but since Twisted depends on it, I've delayed migrating away from it.
> 
> I found that Oliver Palmer has created a new module [pywincffi](https://github.com/opalmer/pywincffi) but it seems that development has stalled, although there is also a very rough PR based upon that work: https://github.com/twisted/twisted/pull/920

Much as I like cffi, I would want to make very sure that we understand the exact build / integration problem that pywin32 is facing before attempting to pursue this particular strategy.  Are these APIs even available under minGW?  If they are, then why won't the existing binary wheels work?  Is the issue merely that the pywin32 maintainers need to be asked to upload a wheel for a different platform tag?

> Any chance new progress could be made on this? 

Be the future that you want to see in the world :).

**Comment by twisted-trac** [comment]
|[<img alt="cas--'s avatar" src="https://avatars.githubusercontent.com/u/606038?s=50" width="50" height="50">](https://github.com/cas--)<a name="note_14"></a>|@cas-- commented|
|-|-|

> Be the future that you want to see in the world :). 

I wish I has the requisite skills here (also I don't use Windows) to move this forward, other than pinging issues ;)

> Is this because pywin32 can't be built against minGW's libraries or C runtime or something? More detail would be helpful. 

Yes. Seems to be something to do with VC exception handling, although that was in 2015 so maybe the code has changed. This is last related issue: https://github.com/msys2/MINGW-packages/issues/751

> it doesn't sound like that's the problem you're having. 

Yes the problem relates to having everything available as MinGW packages so those MSVC compiled wheels are not usable with mingw-w64 Python.

Perhaps your knowledge betters understands the situation than I but certainly having twisted as a [mingw packages](https://github.com/msys2/MINGW-packages) would be super useful.

> which suggests putting all of this stuff in optional, rather than required, dependencies. 

What prevents it being merged? With no wheels on PyPi, need to use this website to install on Windows without compiling: https://www.lfd.uci.edu/~gohlke/pythonlibs/#twisted
